logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageSender.kt

```diff
@@ -26,18 +29,44 @@ class MessageSenderImpl(
             syncManager.waitForSlowSyncToComplete()
 
             // TODO: make this thread safe (per user)
+            getRecipientsAndAttemptSend(conversationId, messageUuid)
+        }
+
+    private suspend fun getRecipientsAndAttemptSend(conversationId: ConversationId, messageUuid: String): Either<CoreFailure, Unit> =
+        suspending {
             conversationRepository.getConversationRecipients(conversationId)
                 .flatMap { recipients ->
                     sessionEstablisher.prepareRecipientsForNewOutgoingMessage(recipients).map { recipients }
                 }.flatMap { recipients ->
                     messageRepository.getMessageById(conversationId, messageUuid).flatMap { message ->
                         messageEnvelopeCreator.createOutgoingEnvelope(recipients, message).flatMap { envelope ->
-                            messageRepository.sendEnvelope(conversationId, envelope)
-                            //TODO("Handle failures too")
+                            trySendingEnvelopeRetryingIfPossible(conversationId, envelope, messageUuid)
                         }.flatMap {
-                            TODO("Mark message as sent")
+                            messageRepository.markMessageAsSent(conversationId, messageUuid)
                         }
                     }
                 }
         }
+
+    private suspend fun trySendingEnvelopeRetryingIfPossible(
+        conversationId: ConversationId,
+        envelope: MessageEnvelope,
+        messageUuid: String,
+    ): Either<CoreFailure, Unit> = suspending {
+        val sendResult = messageRepository.sendEnvelope(conversationId, envelope)
+
+        if (sendResult.isRight()) {
+            return@suspending Either.Right(Unit)
+        }
+
+        when (val failure = sendResult.value) {
+            is SendMessageFailure.ClientsHaveChanged -> {
+                messageSendFailureHandler.handleClientsHaveChangedFailure(failure).flatMap {
+                    getRecipientsAndAttemptSend(conversationId, messageUuid)
+                }
+            }
+            SendMessageFailure.NoNetworkConnection -> Either.Left(CoreFailure.NoNetworkConnection)
+            else -> Either.Left(failure)
```

Comment:
In these cases we should set message status to `FAILED` or try again? I think we need to different strategies for different kind of messages. User created text messages for example we want to fail and notify the user but for non user generated messages we probably want to retry longer.

Reply:
I didn't come up with a full-fledged solution for this yet but it's something I would like to explore when creating the scheduling and retry mechanisms, also when expanding on different message types.

Maybe this part could do that (or delegate to an specialised entity to figure out). Based on the message type and kind of failure we can decide if we:
- Retry
- Give up (marking the message as `FAILED`)
- Just Give Up 

At the moment, this is the current approach:
<img width="200" alt="image" src="https://user-images.githubusercontent.com/9389043/155506259-b731ac84-e09c-4a84-ab8a-294130b0bb86.png">

I'd say we can expand later.

Reply:
Sounds like plan!